hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/DeleteBucketHandler.java

```diff
@@ -18,30 +18,164 @@
+import static org.apache.hadoop.ozone.om.OMConfigKeys.OZONE_OM_SERVICE_IDS_KEY;
+import static org.apache.hadoop.ozone.om.helpers.BucketLayout.OBJECT_STORE;
 
 /**
  * Delete bucket Handler.
  */
 @Command(name = "delete",
-    description = "deletes an empty bucket")
+    description = "deletes a bucket")
 public class DeleteBucketHandler extends BucketHandler {
 
+  @CommandLine.Option(
+      names = {"-r"},
+      description = "Delete bucket recursively"
+  )
+  private boolean bRecursive;
+
+  @CommandLine.Option(
+      names = {"-id", "--om-service-id"},
+      description = "Ozone Manager Service ID"
+  )
+  private String omServiceId;
+
+  @CommandLine.Option(names = {"-y", "--yes"},
+      description = "Continue without interactive user confirmation")
+  private boolean yes;
+
+  private static final int MAX_KEY_DELETE_BATCH_SIZE = 1000;
+
   @Override
   protected void execute(OzoneClient client, OzoneAddress address)
       throws IOException {
 
     String volumeName = address.getVolumeName();
     String bucketName = address.getBucketName();
-
     OzoneVolume vol = client.getObjectStore().getVolume(volumeName);
+
+    if (bRecursive) {
+      Collection<String> serviceIds = getConf().getTrimmedStringCollection(
+          OZONE_OM_SERVICE_IDS_KEY);
+      if (Strings.isNullOrEmpty(omServiceId)) {
+        if (serviceIds.size() > 1) {
+          out().printf("OmServiceID not provided, provide using " +
+              "-id <OM_SERVICE_ID>%n");
+          return;
+        } else if (serviceIds.size() == 1) {
+          // Only one OM service ID configured, we can use that
+          omServiceId = serviceIds.iterator().next();
+        }
+      }
+      if (!yes) {
+        // Ask for user confirmation
+        out().print("This command will delete bucket recursively." +
+            "\nThere is no recovery option after using this command, " +
+            "and no trash for FSO buckets." +
+            "\nEnter 'yes' to proceed': ");
+        out().flush();
+        Scanner scanner = new Scanner(new InputStreamReader(
+            System.in, StandardCharsets.UTF_8));
+        String confirmation = scanner.next().trim().toLowerCase();
+        if (!confirmation.equals("yes")) {
+          out().println("Operation cancelled.");
+          return;
+        }
+      }
+      OzoneBucket bucket = vol.getBucket(bucketName);
+      if (bucket.getBucketLayout().equals(OBJECT_STORE)) {
+        deleteOBSBucketRecursive(vol, bucket);
+      } else {
+        deleteFSBucketRecursive(vol, bucket);
+      }
+      return;
+    }
+    // Delete bucket without recursive
     vol.deleteBucket(bucketName);
+    out().printf("Bucket %s is deleted%n", bucketName);
   }
 
+  /**
+   * Delete OBS bucket recursively.
+   *
+   * @param bucket OzoneBucket
+   */
+  private void deleteOBSBucketRecursive(OzoneVolume vol, OzoneBucket bucket) {
+    ArrayList<String> keys = new ArrayList<>();
+    try {
+      if (!bucket.isLink()) {
```

Comment:
I remember that bucket is link can also perform listKeys, here this treatment is to consider accidental deletion?

Reply:
IMO deleting a bucket link should only delete the link & not the source bucket & its contents.

Reply:
Yes correct same as file system behaviour.